File: text_tool/static/text_tool/fonts.js
// fonts.js — the font catalogue on the browser side.
//
// Fetches /fonts-list (assets/fonts/fonts.json + which files are installed),
// then does the three things every face needs done once:
//   1. @font-face rules for every installed style, so SVG text in
//      font-family "Nimbus Roman" is drawn with the actual URW file — the
//      face MuPDF rendered the scan with — instead of whatever the OS picks;
//   2. the toolbar's font menu (#fabric-font-family), one option per family,
//      MuPDF's own faces first;
//   3. window.FontCatalog for everyone else: has(), familyForPdfName(),
//      select(family, sizePt), metrics(), fileUrl().
//
// Two generic events choose the default face, so no plugin is named here:
//   document:loaded   — the core's declared PDF fonts (pdfFonts) and sampled
//                       body size (sizePt) set the initial selection;
//   typography:detected ({ fontFamily, sizePt, source }) — an analysis plugin
//                       that measured the page (an OCR read, say) overrides it.
(function () {
  const catalog = { families: [], byFamily: new Map(), default: 'Times New Roman', staticBase: '/static/fonts/', ready: null };

  const WEIGHT = { regular: 400, bold: 700, italic: 400, bolditalic: 700 };
  const STYLE = { regular: 'normal', bold: 'normal', italic: 'italic', bolditalic: 'italic' };
  const format = f => /\.otf$/i.test(f) ? 'opentype' : 'truetype';

  function injectFontFaces() {
    let css = '';
    for (const fam of catalog.families)
      for (const [style, file] of Object.entries(fam.files))
        if (fam.present[style])
          css += `@font-face { font-family: "${fam.family}"; src: url("${catalog.staticBase}${file}") format("${format(file)}"); ` +
                 `font-weight: ${WEIGHT[style]}; font-style: ${STYLE[style]}; font-display: block; }\n`;
    let el = document.getElementById('tt-font-faces');
    if (!el) { el = document.createElement('style'); el.id = 'tt-font-faces'; document.head.appendChild(el); }
    el.textContent = css;
  }

  function fillMenu() {
    const sel = document.getElementById('fabric-font-family');
    if (!sel) return;
    const current = sel.value;
    sel.innerHTML = '';
    for (const fam of catalog.families) {
      const opt = document.createElement('option');
      opt.value = fam.family;
      opt.textContent = fam.present.regular ? fam.family : `${fam.family} (not installed)`;
      opt.disabled = !fam.present.regular;
      if (fam.class === 'mupdf') opt.title = fam.note || 'one of MuPDF\'s own faces';
      sel.appendChild(opt);
    }
    if (current && catalog.byFamily.has(current)) sel.value = current;
    else sel.value = catalog.default;
  }

  function has(family) { return catalog.byFamily.has(family); }

  // 'ABCDEF+TimesNewRomanPSMT-Bold' → 'Times New Roman'; longest alias wins
  function familyForPdfName(name) {
    if (!name) return null;
    const key = String(name).split('+').pop().replace(/[\s,-]/g, '').toLowerCase();
    let best = null, bestLen = 0;
    for (const fam of catalog.families)
      for (const alias of fam.pdfNames || []) {
        const a = alias.replace(/[\s,-]/g, '').toLowerCase();
        if (key.startsWith(a) && a.length > bestLen) { best = fam.family; bestLen = a.length; }
      }
    return best;
  }

  // set the toolbar's font (and size) — the default for the next added box.
  // source ranks the claim for this document: 'detected' (a plugin measured
  // the page's face from its pixels) outranks 'declared' (the PDF's font
  // names) and 'layer' (the text layer's most used face), which are what a
  // producer wrote, not what the page shows; a weaker claim arriving after a
  // stronger one is ignored, so the reader's face is not undone by the
  // layer's spans landing later. Unranked calls (the user's own choice) win.
  const RANK = { layer: 1, declared: 1, detected: 2 };
  let chosen = 0;
  function select(family, sizePt, source) {
    const rank = source ? (RANK[source] || 1) : 3;
    if (rank < chosen) return false;
    if (source) chosen = rank;
    const sel = document.getElementById('fabric-font-family');
    if (sel && family) {
      if (!Array.from(sel.options).some(o => o.value === family)) {
        const opt = document.createElement('option');
        opt.value = family; opt.textContent = `${family} (not installed)`;
        sel.appendChild(opt);
      }
      sel.value = family;
      if (typeof textOptions !== 'undefined') textOptions.fontFamily = family;
    }
    const size = document.getElementById('fabric-font-size');
    if (size && sizePt > 0) size.value = Math.round(sizePt * 100) / 100;
    return true;
  }

  catalog.ready = (async () => {
    try {
      const r = await fetch('/fonts-list', { cache: 'no-store' });
      const data = await r.json();
      catalog.families = Array.isArray(data) ? [] : (data.families || []);
      catalog.default = data.default || catalog.default;
      catalog.staticBase = data.static || catalog.staticBase;
      catalog.byFamily = new Map(catalog.families.map(f => [f.family, f]));
      injectFontFaces();
      fillMenu();
    } catch (e) {
      console.warn('font catalogue unavailable:', e);
    }
    return catalog;
  })();

  PDFHooks.on('document:loaded', async e => {
    chosen = 0;                                   // a new document: every claim is open again
    await catalog.ready;
    const declared = (e?.pdfFonts || []).map(familyForPdfName).find(Boolean);
    select(declared || catalog.default, e?.sizePt, 'declared');
  });
  PDFHooks.on('typography:detected', async e => {
    await catalog.ready;
    if (e?.fontFamily) select(e.fontFamily, e.sizePt, 'detected');
  });

  // The face's own advances and kern pairs at a pixel size, from HarfBuzz
  // (/font-metrics): {space, adv: {ch: px}, kern: {pair: px}} or null. A
  // plugin that learned from a page's pens whether its producer kerned lays
  // pairs the page never wrote with this table.
  const metricsCache = new Map();
  function metrics(family, bold, italic, sizePx) {
    const key = `${family}|${bold ? 1 : 0}|${italic ? 1 : 0}|${+sizePx}`;
    if (!metricsCache.has(key)) metricsCache.set(key, (async () => {
      try {
        const r = await fetch(`/font-metrics?family=${encodeURIComponent(family)}&bold=${bold ? 1 : 0}&italic=${italic ? 1 : 0}&size_px=${+sizePx}`);
        return r.ok ? await r.json() : null;
      } catch { return null; }
    })());
    return metricsCache.get(key);
  }

  // The installed file of one style of a family, as a URL — what a renderer
  // that draws glyphs itself loads. null when that style is not installed (a
  // family without a bold italic has none to draw; nothing is synthesized).
  function fileUrl(family, bold, italic) {
    const fam = catalog.byFamily.get(family);
    if (!fam) return null;
    const style = bold && italic ? 'bolditalic' : bold ? 'bold' : italic ? 'italic' : 'regular';
    return fam.files?.[style] && fam.present?.[style] ? catalog.staticBase + fam.files[style] : null;
  }

  window.FontCatalog = { has, familyForPdfName, select, metrics, fileUrl,
    families: () => catalog.families, get ready() { return catalog.ready; }, get default() { return catalog.default; } };
})();

File: text_tool/logic/fonts.py
"""The font catalogue — assets/fonts/fonts.json read once, resolved everywhere.

One list of faces for the whole app: the toolbar's font menu, the @font-face
rules the browser draws vector text with, and the file HarfBuzz measures with
all come from here, so "Nimbus Roman bold" means the same file in every
place. Family names match tol0's generated set-fonts table (the face each
OCR glyph set was rendered from), which is what lets an OCR line name its
face and the text tool select it.

`python manage.py fonts_setup` builds the files the catalogue names.
"""
import json
import os
from pathlib import Path

FONT_DIR = Path(__file__).resolve().parents[2] / 'assets' / 'fonts'
CATALOG_PATH = FONT_DIR / 'fonts.json'

STYLE_KEYS = ('regular', 'bold', 'italic', 'bolditalic')
FALLBACK_FAMILY = 'Times New Roman'

_cache = {'mtime': None, 'data': None}


def catalog():
    """The parsed catalogue (cached by file mtime): {'families': [...]}."""
    try:
        mtime = os.path.getmtime(CATALOG_PATH)
    except OSError:
        return {'families': []}
    if _cache['mtime'] != mtime:
        with open(CATALOG_PATH, encoding='utf-8') as f:
            data = json.load(f)
        _cache.update(mtime=mtime, data=data)
    return _cache['data']


def families():
    """Catalogue entries with a `present` map: which style files exist on disk."""
    out = []
    for fam in catalog().get('families', []):
        files = fam.get('files', {})
        out.append({
            'family': fam['family'],
            'class': fam.get('class', ''),
            'note': fam.get('note', ''),
            'pdfNames': fam.get('pdfNames', []),
            'files': files,
            'present': {k: (FONT_DIR / v).is_file() for k, v in files.items()},
        })
    return out


def _entry(family):
    key = (family or '').strip().lower()
    for fam in catalog().get('families', []):
        if fam['family'].lower() == key:
            return fam
    return None


def style_key(bold=False, italic=False):
    return 'bolditalic' if bold and italic else 'bold' if bold else 'italic' if italic else 'regular'


def resolve(family, bold=False, italic=False):
    """Path of the file for (family, bold, italic), falling back through the
    family's styles (bold italic → bold → italic → regular) and finally to the
    fallback family's regular. None only when nothing at all is installed."""
    fam = _entry(family) or _entry(FALLBACK_FAMILY)
    if not fam:
        return None
    files = fam.get('files', {})
    want = style_key(bold, italic)
    chain = {
        'bolditalic': ('bolditalic', 'bold', 'italic', 'regular'),
        'bold': ('bold', 'regular'),
        'italic': ('italic', 'regular'),
        'regular': ('regular',),
    }[want]
    for key in chain:
        name = files.get(key)
        if name and (FONT_DIR / name).is_file():
            return str(FONT_DIR / name)
    if fam['family'] != FALLBACK_FAMILY:
        return resolve(FALLBACK_FAMILY, bold, italic)
    return None


def resolve_file(name):
    """Legacy file-name API (`font: 'times.ttf'`): the file in assets/fonts, or None."""
    if not name:
        return None
    for candidate in (Path(name), FONT_DIR / name, FONT_DIR / (name + '.ttf')):
        if candidate.is_file():
            return str(candidate)
    return None


def family_for_pdf_name(pdf_name):
    """A PDF BaseFont name (e.g. 'ABCDEF+TimesNewRomanPSMT-Bold') → catalogue
    family, or None. Longest alias wins so 'TimesNewRoman' beats 'Times'."""
    if not pdf_name:
        return None
    key = pdf_name.split('+')[-1].replace(' ', '').replace('-', '').replace(',', '').lower()
    best, best_len = None, 0
    for fam in catalog().get('families', []):
        for alias in fam.get('pdfNames', []):
            a = alias.replace(' ', '').replace('-', '').lower()
            if key.startswith(a) and len(a) > best_len:
                best, best_len = fam['family'], len(a)
    return best


def default_family():
    fam = _entry(FALLBACK_FAMILY)
    if fam and resolve(fam['family']):
        return fam['family']
    for f in families():
        if f['present'].get('regular'):
            return f['family']
    return FALLBACK_FAMILY
